# Connect to sql database and fix display option

In [2]:
import prettytable
if not hasattr(prettytable, "DEFAULT"):
    prettytable.DEFAULT = prettytable.PLAIN_COLUMNS

C:\Users\Niklas\AppData\Local\Temp\ipykernel_14452\3990275676.py:2: DeprecationWarning: the 'DEFAULT' constant is deprecated, use the 'TableStyle' enum instead
  if not hasattr(prettytable, "DEFAULT"):


In [3]:
%load_ext sql

In [4]:
%sql postgresql://postgres:ajkhsj7@localhost:5432/World_Bank

In [5]:
%sql SELECT version();

 * postgresql://postgres:***@localhost:5432/World_Bank
1 rows affected.


[('PostgreSQL 17.5 on x86_64-windows, compiled by msvc-19.44.35209, 64-bit',)]

# Investigate the Data (Basic Insights)

## Count the number of distinct countries in the dataset
- Note: Some countries do not report all indicators, so null values may be present in some series (will be seen in advanced analysis).

In [6]:
%%sql 
SELECT COUNT(DISTINCT country_code) AS total_distinct_countries
FROM international_debt;

 * postgresql://postgres:***@localhost:5432/World_Bank
1 rows affected.


[(120,)]

## Investigating the definition and structure of key debt ratios

In [7]:
%%sql 
SELECT country_name, series_name, series_code, value, year, counterpart_area_name
FROM international_debt
WHERE series_code = 'DT.DOD.DECT.GN.ZS'
ORDER BY country_name
Limit 10;



 * postgresql://postgres:***@localhost:5432/World_Bank
10 rows affected.


[('Afghanistan', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('13.9496520351375'), 2019, 'World'),
 ('Afghanistan', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('24.7857399527695'), 2021, 'World'),
 ('Afghanistan', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('15.1208585304773'), 2020, 'World'),
 ('Afghanistan', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('23.3368489769108'), 2022, 'World'),
 ('Afghanistan', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('19.7980988255309'), 2023, 'World'),
 ('Albania', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('60.2870294294132'), 2019, 'World'),
 ('Albania', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('69.8890319190523'), 2020, 'World'),
 ('Albania', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('62.1722783747637'), 2021, 'World'),
 ('Albania', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('55.7646367740804'), 2022, 'World'),
 ('Albania', 'External debt stocks (% of GNI)', 'DT.DOD.DECT.GN.ZS', Decimal('48.8846201182997'), 2023, 'World')]

In [8]:
%%sql
SELECT country_name, series_name, series_code, value, year, counterpart_area_name
FROM international_debt
WHERE series_code = 'DT.DOD.DECT.PC.CD'
ORDER BY country_name
Limit 10;

 * postgresql://postgres:***@localhost:5432/World_Bank
10 rows affected.


[('Afghanistan', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('70.3105834985048'), 2019, 'World'),
 ('Afghanistan', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('88.8936928849633'), 2021, 'World'),
 ('Afghanistan', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('77.8129449556386'), 2020, 'World'),
 ('Afghanistan', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('83.6210959814969'), 2022, 'World'),
 ('Afghanistan', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('82.6953742538764'), 2023, 'World'),
 ('Albania', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('3249.35135563808'), 2019, 'World'),
 ('Albania', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('3691.95997383934'), 2020, 'World'),
 ('Albania', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('3931.74451787659'), 2021, 'World'),
 ('Albania', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('3751.38948809604'), 2022, 'World'),
 ('Albania', 'Total external debt per capita (US$)', 'DT.DOD.DECT.PC.CD', Decimal('4138.40881571261'), 2023, 'World')]

In [9]:
%%sql
SELECT country_name, series_name, series_code, value, year, counterpart_area_name
FROM international_debt
WHERE series_code = 'DT.DOD.DECT.CD'
ORDER BY country_name
Limit 10;

 * postgresql://postgres:***@localhost:5432/World_Bank
10 rows affected.


[('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('4701000'), 2019, 'Bulgaria'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('276923268.4'), 2019, 'International Monetary Fund'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('10148000'), 2019, 'Iran Islamic Republic Of'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('41046411.8'), 2019, 'Islamic Dev. Bank'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('23928420'), 2019, 'Italy'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('22722323'), 2019, 'Kuwait'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('1550000'), 2019, 'OPEC Fund for International Dev.'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('154590000'), 2019, 'Other Multiple Lenders'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('1134675000'), 2019, 'Russian Federation'),
 ('Afghanistan', 'External debt stocks total (DOD current US$)', 'DT.DOD.DECT.CD', Decimal('573753000'), 2019, 'Asian Dev. Bank')]

Observation:
- Debt-to-GNI ratios are reported as single values per year, always using the World Bank as the counterpart.
- Therefore, no aggregation is required across counterpart areas for this indicator, unlike for total external debt.


## Countries with highest external debt in 2019/2023

In [10]:
%%sql
SELECT country_name, 
	   Round(SUM(value) FILTER (WHERE year = 2019)/ 1000000000,2) ||'B' AS total_debt_2019,
	   Round(SUM(value) FILTER (WHERE year = 2023)/ 1000000000,2) ||'B' AS total_debt_2023
FROM international_debt
WHERE series_code = 'DT.DOD.DECT.CD' 
GROUP BY country_name
ORDER BY SUM(value) FILTER (WHERE year = 2019) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/World_Bank
10 rows affected.


[('China', '4228.32B', '4840.42B'),
 ('Mexico', '1234.87B', '1191.84B'),
 ('Brazil', '1137.36B', '1213.80B'),
 ('India', '1121.98B', '1293.27B'),
 ('Turkiye', '829.73B', '998.23B'),
 ('Indonesia', '804.75B', '812.11B'),
 ('Argentina', '598.71B', '532.33B'),
 ('South Africa', '381.28B', '327.82B'),
 ('Thailand', '337.20B', '387.25B'),
 ('Kazakhstan', '318.56B', '325.63B')]

Insight:
- Very mixed development. 5 countries increased debt (China, Brazil, India, Turkiye and Thailand) 
- 3 countries experienced only slight changes in their debt (Kazakhstan, Mexico and Indonesia)
- 2 reduced debt (South Africa, Argentina)

## Biggest lenders in 2019 and 2023

In [11]:
%%sql
SELECT counterpart_area_name,
	   Round(SUM(value) FILTER (WHERE year = 2019)/ 1000000,2) ||'M' AS biggest_lender_2019,
	   Round(SUM(value) FILTER (WHERE year = 2023)/ 1000000,2) ||'M' AS biggest_lender_2023
FROM international_debt
WHERE series_code = 'DT.DOD.DECT.CD' 
GROUP BY counterpart_area_name
ORDER BY sum(value) FILTER (WHERE year = 2019) DESC
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/World_Bank
10 rows affected.


[('World', '7755392.50M', '8838932.09M'),
 ('Other Multiple Lenders', '4079845.61M', '4435016.09M'),
 ('Bondholders', '1978811.26M', '2155490.17M'),
 ('China', '195574.83M', '193041.24M'),
 ('World Bank-IBRD', '181913.22M', '240162.62M'),
 ('International Monetary Fund', '163649.57M', '383111.69M'),
 ('World Bank-IDA', '159589.63M', '196663.35M'),
 ('Multiple Lenders', '146233.33M', '175547.30M'),
 ('Japan', '123458.35M', '120965.17M'),
 ('Asian Dev. Bank', '108649.77M', '145827.42M')]

Observations:
- As expected, the World Bank is the largest external lender.
- Only two countries/orginaizations which didn't lend more in 2023 (China and Japan).

# Advanced Analysis

## Investigating and comparing debt per capita and % of GNI
- (High debt per capita + high % of GNI = potentially risky)

In [12]:
%%sql
SELECT country_name, year, 
       ROUND(SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END)/ 1000000000,2) ||'B' AS total_debt,
	   ROUND(MAX(CASE WHEN series_code = 'DT.DOD.DECT.GN.ZS' THEN value END)) || '%' AS debt_gni_ratio,
	   ROUND(MAX(CASE WHEN series_code = 'DT.DOD.DECT.PC.CD' THEN value END),2) AS debt_per_capita
FROM international_debt 
WHERE series_code IN ('DT.DOD.DECT.CD','DT.DOD.DECT.PC.CD', 'DT.DOD.DECT.GN.ZS')
	  AND year = 2019
GROUP BY country_name, year 
HAVING MAX (CASE WHEN series_code = 'DT.DOD.DECT.GN.ZS' THEN value END) IS NOT NULL
ORDER BY MAX(CASE WHEN series_code = 'DT.DOD.DECT.PC.CD' THEN value END) DESC
limit 10;

 * postgresql://postgres:***@localhost:5432/World_Bank
10 rows affected.


[('Montenegro', 2019, '16.87B', '151%', Decimal('13582.02')),
 ('Mongolia', 2019, '66.46B', '264%', Decimal('10192.99')),
 ('Mauritius', 2019, '24.71B', '82%', Decimal('9764.01')),
 ('Kazakhstan', 2019, '318.56B', '100%', Decimal('8291.79')),
 ('Dominica', 2019, '0.92B', '76%', Decimal('6779.77')),
 ('Argentina', 2019, '598.71B', '70%', Decimal('6656.24')),
 ('Suriname', 2019, '7.86B', '109%', Decimal('6483.56')),
 ('Costa Rica', 2019, '59.46B', '49%', Decimal('5960.86')),
 ('Maldives', 2019, '5.35B', '52%', Decimal('5484.19')),
 ('Jamaica', 2019, '28.87B', '94%', Decimal('5123.37'))]

Observations:
- Mongolia and Montenegro appear potentially risky with high debt compared to GNI and per capita



## Comparing debt-to-GNI ratios across years (2019 vs 2023)

In [13]:
%%sql
SELECT country_name, 
	   Round(MAX(value) FILTER (WHERE year = 2019)) AS debt_gni_ratio_2019,
	   Round(MAX(value) FILTER (WHERE year = 2023)) AS debt_gni_ratio_2023
FROM international_debt
WHERE series_code = 'DT.DOD.DECT.GN.ZS'
GROUP BY country_name
HAVING  AVG(value) FILTER (WHERE year = 2019) >= 100
ORDER BY debt_gni_ratio_2019 DESC;

 * postgresql://postgres:***@localhost:5432/World_Bank
9 rows affected.


[('Mozambique', Decimal('356'), Decimal('350')),
 ('Mongolia', Decimal('264'), Decimal('190')),
 ('Montenegro', Decimal('151'), Decimal('113')),
 ('Zambia', Decimal('128'), Decimal('110')),
 ('Nicaragua', Decimal('111'), Decimal('90')),
 ('Georgia', Decimal('111'), Decimal('86')),
 ('Lao PDR', Decimal('109'), Decimal('139')),
 ('Suriname', Decimal('109'), Decimal('129')),
 ('Kazakhstan', Decimal('100'), Decimal('69'))]

Observations: 9 countries have debt exceeding their annual GNI (2019)
- Mozambique and Mongolia stand out with high ratios (Mozambique was not in above query as i sorted by debt per capita
- Many countries reduced debt significantly — Mongolia improved, but Georgia and Lao PDR saw increases

## Breakdown of public vs private debt in 2019
- Focus on countries where public debt is >60% of total debt

In [14]:
%%sql
SELECT
    country_name,
    year,
    SUM(CASE WHEN series_code = 'DT.DOD.DPPG.CD' THEN value END) AS public_debt,
    SUM(CASE WHEN series_code = 'DT.DOD.DPNG.CD' THEN value END) AS private_debt,
    ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DPPG.CD' THEN value END) /
                 SUM(CASE WHEN series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD') THEN value END), 2)
        AS pct_public,
    ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DPNG.CD' THEN value END) /
                 SUM(CASE WHEN series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD') THEN value END), 2)
        AS pct_private
FROM international_debt
WHERE series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD')
  AND year = 2019
GROUP BY country_name, year
HAVING ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DPPG.CD' THEN value END) /
                 SUM(CASE WHEN series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD') THEN value END), 2) > 60
ORDER BY pct_public DESC;

 * postgresql://postgres:***@localhost:5432/World_Bank
81 rows affected.


[('Ethiopia', 2019, Decimal('59153623171.6'), None, Decimal('100.00'), None),
 ('Sao Tome and Principe', 2019, Decimal('481398295.4'), None, Decimal('100.00'), None),
 ('Djibouti', 2019, Decimal('4291076125.6'), None, Decimal('100.00'), None),
 ('St. Vincent and the Grenadines', 2019, Decimal('642570986.8'), None, Decimal('100.00'), None),
 ('Sierra Leone', 2019, Decimal('2481292356.6'), None, Decimal('100.00'), None),
 ('Mali', 2019, Decimal('8895110678.4'), None, Decimal('100.00'), None),
 ('Sudan', 2019, Decimal('33588338231.4'), None, Decimal('100.00'), None),
 ('Cabo Verde', 2019, Decimal('3619212761.0'), None, Decimal('100.00'), None),
 ('Guinea', 2019, Decimal('5744935443.4'), None, Decimal('100.00'), None),
 ('Comoros', 2019, Decimal('476186319.0'), None, Decimal('100.00'), None),
 ('Congo Dem. Rep.', 2019, Decimal('9977961675.0'), None, Decimal('100.00'), None),
 ('Gabon', 2019, Decimal('12408046637.0'), None, Decimal('100.00'), None),
 ('Niger', 2019, Decimal('6307304246.8'), None, Decimal('100.00'), None),
 ('Vanuatu', 2019, Decimal('690935143.2'), None, Decimal('100.00'), None),
 ('Malawi', 2019, Decimal('4039216472.4'), None, Decimal('100.00'), None),
 ('Burundi', 2019, Decimal('957928064.9'), None, Decimal('100.00'), None),
 ('Togo', 2019, Decimal('2705965616.4'), None, Decimal('100.00'), None),
 ('Mauritania', 2019, Decimal('8108806994.0'), None, Decimal('100.00'), None),
 ('Eritrea', 2019, Decimal('1451297473.4'), None, Decimal('100.00'), None),
 ('Congo Rep.', 2019, Decimal('15867789041.0'), None, Decimal('100.00'), None),
 ('Chad', 2019, Decimal('5439773532.2'), None, Decimal('100.00'), None),
 ('Timor-Leste', 2019, Decimal('382434387.0'), None, Decimal('100.00'), None),
 ('Central African Republic', 2019, Decimal('841508299.4'), None, Decimal('100.00'), None),
 ('Iraq', 2019, Decimal('45414864136.2'), None, Decimal('100.00'), None),
 ('St. Lucia', 2019, Decimal('1084461052.6'), None, Decimal('100.00'), None),
 ('Somalia', 2019, Decimal('3827364353.6'), None, Decimal('100.00'), None),
 ('Syrian Arab Republic', 2019, Decimal('7269731593.2'), None, Decimal('100.00'), None),
 ('Haiti', 2019, Decimal('4024724249.6'), Decimal('0'), Decimal('100.00'), Decimal('0.00')),
 ('Gambia The', 2019, Decimal('1491779465.6'), None, Decimal('100.00'), None),
 ('Tonga', 2019, Decimal('355093877.0'), None, Decimal('100.00'), None),
 ('Yemen Rep.', 2019, Decimal('12376067315.2'), None, Decimal('100.00'), None),
 ('Grenada', 2019, Decimal('992931588.8'), None, Decimal('100.00'), None),
 ('Myanmar', 2019, Decimal('21496945474.4'), Decimal('55690000'), Decimal('99.74'), Decimal('0.26')),
 ('Egypt Arab Rep.', 2019, Decimal('180177904568.7'), Decimal('720476000'), Decimal('99.60'), Decimal('0.40')),
 ('Samoa', 2019, Decimal('775208219.4'), Decimal('3986000'), Decimal('99.49'), Decimal('0.51')),
 ('Turkmenistan', 2019, Decimal('13021069891.2'), Decimal('86706258.8'), Decimal('99.34'), Decimal('0.66')),
 ('Botswana', 2019, Decimal('2666776567.6'), Decimal('24269682'), Decimal('99.10'), Decimal('0.90')),
 ('Afghanistan', 2019, Decimal('3887215376.2'), Decimal('42586000'), Decimal('98.92'), Decimal('1.08')),
 ('Bhutan', 2019, Decimal('5432585978.4'), Decimal('102908000'), Decimal('98.14'), Decimal('1.86')),
 ('Nepal', 2019, Decimal('11694351714.8'), Decimal('236542000'), Decimal('98.02'), Decimal('1.98')),
 ('Kenya', 2019, Decimal('61625881460.8'), Decimal('1634567873.6'), Decimal('97.42'), Decimal('2.58')),
 ('Madagascar', 2019, Decimal('6233096490.6'), Decimal('202712000'), Decimal('96.85'), Decimal('3.15')),
 ('Guinea-Bissau', 2019, Decimal('1104384835.4'), Decimal('47140000'), Decimal('95.91'), Decimal('4.09')),
 ('Maldives', 2019, Decimal('4453026797.4'), Decimal('207444000'), Decimal('95.55'), Decimal('4.45')),
 ('Belize', 2019, Decimal('2570861611.4'), Decimal('126052000'), Decimal('95.33'), Decimal('4.67')),
 ('Liberia', 2019, Decimal('1643732817.4'), Decimal('97066455.6'), Decimal('94.42'), Decimal('5.58')),
 ('Dominica', 2019, D

Insight: 81 of 120 countries have a public debt share above 60%

## Opposite of the above — countries with <40% public debt
- Highlights where private, non-guaranteed debt is dominant

In [15]:
%%sql
SELECT
    country_name,
    year,
    SUM(CASE WHEN series_code = 'DT.DOD.DPPG.CD' THEN value END) AS public_debt,
    SUM(CASE WHEN series_code = 'DT.DOD.DPNG.CD' THEN value END) AS private_debt,
    ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DPPG.CD' THEN value END) /
      SUM(CASE WHEN series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD') THEN value END), 2) AS pct_public,
    ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DPNG.CD' THEN value END) /
      SUM(CASE WHEN series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD') THEN value END), 2)AS pct_private
FROM international_debt
WHERE series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD')
  AND year = 2019
GROUP BY country_name, year
HAVING ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DPPG.CD' THEN value END) /
       SUM(CASE WHEN series_code IN ('DT.DOD.DPPG.CD', 'DT.DOD.DPNG.CD') THEN value END), 2) < 40
ORDER BY pct_public DESC;

 * postgresql://postgres:***@localhost:5432/World_Bank
14 rows affected.


[('Brazil', 2019, Decimal('387541291068.4'), Decimal('583473920000'), Decimal('39.91'), Decimal('60.09')),
 ('Solomon Islands', 2019, Decimal('196150181.0'), Decimal('311510000'), Decimal('38.64'), Decimal('61.36')),
 ('Burkina Faso', 2019, Decimal('6748234486.8'), Decimal('10950718000'), Decimal('38.13'), Decimal('61.87')),
 ('China', 2019, Decimal('636128537816.2'), Decimal('1162240296845.8'), Decimal('35.37'), Decimal('64.63')),
 ('Thailand', 2019, Decimal('81768242997.8'), Decimal('151119224000'), Decimal('35.11'), Decimal('64.89')),
 ('Peru', 2019, Decimal('37210530660.4'), Decimal('72462142000'), Decimal('33.93'), Decimal('66.07')),
 ('Mongolia', 2019, Decimal('20014522376.8'), Decimal('40096340000'), Decimal('33.30'), Decimal('66.70')),
 ('Iran Islamic Rep.', 2019, Decimal('740164276.6'), Decimal('1829372000'), Decimal('28.81'), Decimal('71.19')),
 ('Moldova', 2019, Decimal('2846075080.8'), Decimal('7219456000'), Decimal('28.28'), Decimal('71.72')),
 ('Kosovo', 2019, Decimal('639498870.8'), Decimal('2074982000'), Decimal('23.56'), Decimal('76.44')),
 ('Papua New Guinea', 2019, Decimal('8053116759.8'), Decimal('27202000000'), Decimal('22.84'), Decimal('77.16')),
 ('Mauritius', 2019, Decimal('2762915361.6'), Decimal('10259034000'), Decimal('21.22'), Decimal('78.78')),
 ('Mozambique', 2019, Decimal('20764351169.4'), Decimal('83806938000'), Decimal('19.86'), Decimal('80.14')),
 ('Kazakhstan', 2019, Decimal('50003466603.2'), Decimal('249557038000'), Decimal('16.69'), Decimal('83.31'))]

- Mozambique and Mongolia again appear — most of their debt is private

## Short-term external debt ratios
- A high percentage can be risky due to refinancing pressure

In [16]:
%%sql
SELECT
    country_name,
    year,
	ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DSTC.CD' THEN value END) /
      SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END), 2) AS short_term_ratio_pct,
    ROUND(SUM(CASE WHEN series_code = 'DT.DOD.DSTC.CD' THEN value END)/1000000000,2) ||'B' AS short_term_debt,
    ROUND(SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END)/1000000000,2) ||'B' AS total_external_debt    
FROM international_debt
WHERE series_code IN ('DT.DOD.DSTC.CD', 'DT.DOD.DECT.CD')
  AND year = 2019
GROUP BY country_name, year
HAVING ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DSTC.CD' THEN value END) /
              SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END), 0) > 30
ORDER BY short_term_ratio_pct DESC;

 * postgresql://postgres:***@localhost:5432/World_Bank
8 rows affected.


[('Somalia', 2019, Decimal('57.87'), '5.86B', '10.12B'),
 ('China', 2019, Decimal('57.01'), '2410.62B', '4228.32B'),
 ('Mauritius', 2019, Decimal('46.23'), '11.42B', '24.71B'),
 ('Algeria', 2019, Decimal('41.43'), '4.53B', '10.93B'),
 ('Dominica', 2019, Decimal('37.21'), '0.34B', '0.92B'),
 ('Jordan', 2019, Decimal('36.72'), '24.59B', '66.99B'),
 ('Kosovo', 2019, Decimal('34.55'), '1.67B', '4.83B'),
 ('Iran Islamic Rep.', 2019, Decimal('32.92'), '3.20B', '9.71B')]

- Observation: Mozambique and Mongolia, despite high overall debt, have low short-term debt — this reduces near-term risk

## Composite debt risk score
- Creating a own debt risk score with 3 criteria:
    -   1. Debt-to-GNI > 60%
    -   2. Debt per capita > $5000
    -   3. Short-term debt > 30% of total external debt
- Risk Score: 0–3 (number of criteria met)
- Filtering for countries with a risk score of 2 or 3

In [17]:
%%sql
SELECT country_name,
	   year,
	   ROUND(MAX(CASE WHEN series_code = 'DT.DOD.DECT.GN.ZS' THEN value END),2)|| '%' AS debt_gni_ratio,
	   ROUND(MAX(CASE WHEN series_code = 'DT.DOD.DECT.PC.CD' THEN value END)) AS debt_per_capita,
	   ROUND(100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DSTC.CD' THEN value END) /
      	SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END), 2) || '%' AS short_term_ratio_pct,
		CASE WHEN MAX(CASE WHEN series_code = 'DT.DOD.DECT.GN.ZS' THEN value END) > 60 THEN 1 ELSE 0 END +
		CASE WHEN MAX(CASE WHEN series_code = 'DT.DOD.DECT.PC.CD' THEN value END) > 5000 THEN 1 ELSE 0 END +
		CASE WHEN 100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DSTC.CD' THEN value END) /
      	SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END) > 30 THEN 1 ELSE 0 END 
		  AS risk_score
FROM international_debt
WHERE series_code IN  ('DT.DOD.DECT.GN.ZS', 'DT.DOD.DECT.PC.CD', 'DT.DOD.DSTC.CD', 'DT.DOD.DECT.CD')
	  AND year = 2019 
GROUP BY country_name , year
HAVING CASE WHEN MAX(CASE WHEN series_code = 'DT.DOD.DECT.GN.ZS' THEN value END) > 60 THEN 1 ELSE 0 END +
		CASE WHEN MAX(CASE WHEN series_code = 'DT.DOD.DECT.PC.CD' THEN value END) > 5000 THEN 1 ELSE 0 END +
		CASE WHEN 100.0 * SUM(CASE WHEN series_code = 'DT.DOD.DSTC.CD' THEN value END) /
      	SUM(CASE WHEN series_code = 'DT.DOD.DECT.CD' THEN value END) > 30 THEN 1 ELSE 0 END  >= 2
ORDER BY risk_score DESC;

 * postgresql://postgres:***@localhost:5432/World_Bank
11 rows affected.


[('Mauritius', 2019, '81.57%', Decimal('9764'), '46.23%', 3),
 ('Dominica', 2019, '76.44%', Decimal('6780'), '37.21%', 3),
 ('Georgia', 2019, '111.37%', Decimal('5042'), '12.13%', 2),
 ('Argentina', 2019, '69.59%', Decimal('6656'), '22.39%', 2),
 ('Suriname', 2019, '108.98%', Decimal('6484'), '9.15%', 2),
 ('Kazakhstan', 2019, '100.19%', Decimal('8292'), '5.66%', 2),
 ('Jamaica', 2019, '93.99%', Decimal('5123'), '15.46%', 2),
 ('Mongolia', 2019, '263.63%', Decimal('10193'), '8.69%', 2),
 ('Montenegro', 2019, '150.75%', Decimal('13582'), '3.29%', 2),
 ('Somalia', 2019, '62.35%', Decimal('335'), '57.87%', 2),
 ('Jordan', 2019, '75.25%', Decimal('3138'), '36.72%', 2)]

- Insight: 9 countries meet 2 risk criteria; 2 countries meet all 3, indicating highest concern